gtn/functions/compose.cpp

```diff
@@ -450,3 +456,936 @@ Graph compose(
+      graphDP1.start[dstNodePair.first] && graphDP2.start[dstNodePair.second],
+      graphDP1.accept[dstNodePair.first] &&
+          graphDP2.accept[dstNodePair.second]);
+
+  return std::make_tuple(srcNodeStartAndAccept, dstNodeStartAndAccept);
+}
+
+} // namespace
+
+// Convert from AOS to SOA
+// Assumption: nodes are numbered [0..graph.numNodes())
+//           : arcs are numbered [0..graph.numArcs())
+GraphDataParallel convertToDataParallel(const Graph& graph) {
+  GraphDataParallel graphDP;
+
+  assert(graphDP.accept.empty());
+  assert(graphDP.start.empty());
+
+  // Safe since sizes are asserted to 0
+  graphDP.accept.resize(graph.numNodes(), false);
+  graphDP.start.resize(graph.numNodes(), false);
+
+  graphDP.inArcOffset.resize(graph.numNodes());
+  graphDP.outArcOffset.resize(graph.numNodes());
+
+  graphDP.inArcs.resize(graph.numArcs());
+  graphDP.outArcs.resize(graph.numArcs());
+
+  graphDP.ilabels.resize(graph.numArcs());
+  graphDP.olabels.resize(graph.numArcs());
+  graphDP.srcNodes.resize(graph.numArcs());
+  graphDP.dstNodes.resize(graph.numArcs());
+  graphDP.weights.resize(graph.numArcs());
+
+  for (auto i : graph.accept()) {
+    assert(i >= 0);
+    assert(i < graph.numNodes());
+    graphDP.accept[i] = true;
+  }
+
+  for (auto i : graph.start()) {
+    assert(i >= 0);
+    assert(i < graph.numNodes());
+    graphDP.start[i] = true;
+  }
+
+  for (int i = 0; i < graph.numNodes(); ++i) {
+    graphDP.inArcOffset[i] = graph.numIn(i);
+    graphDP.outArcOffset[i] = graph.numOut(i);
+  }
+
+  // Scan of offsets
+  const int totalInArcs = prefixSumScan(graphDP.inArcOffset, false);
+  const int totalOutArcs = prefixSumScan(graphDP.outArcOffset, false);
+  assert(totalInArcs == totalOutArcs);
+  assert(totalInArcs == graph.numArcs());
+
+  for (int i = 0; i < graph.numNodes(); ++i) {
+    int offset = graphDP.outArcOffset[i];
+
+    for (auto j : graph.out(i)) {
+      assert(j >= 0);
+      assert(j < graph.numArcs());
+      graphDP.outArcs[offset] = j;
+      offset++;
+
+      graphDP.ilabels[j] = graph.ilabel(j);
+      graphDP.olabels[j] = graph.olabel(j);
+      graphDP.srcNodes[j] = graph.srcNode(j);
+      graphDP.dstNodes[j] = graph.dstNode(j);
+      graphDP.weights[j] = graph.weight(j);
+    }
+  }
+
+  for (int i = 0; i < graph.numNodes(); ++i) {
+    int offset = graphDP.inArcOffset[i];
+
+    for (auto j : graph.in(i)) {
+      assert(j >= 0);
+      assert(j < graph.numArcs());
+      graphDP.inArcs[offset] = j;
+      offset++;
+    }
+  }
+
+  return graphDP;
+}
+
+// Convert from SOA to AOS
+// The Graph is supposed to have no nodes and arcs and only supposed to have
+// inputs set
+// Assumption: nodes are numbered [0..graph.numNodes())
+//           : arcs are numbered [0..graph.numArcs())
+void convertFromDataParallel(const GraphDataParallel& graphDP, Graph& graph) {
+  // Some sanity checks
+  assert(graph.numArcs() == 0);
+  assert(graph.numNodes() == 0);
+  assert(graph.inputs().size() == 2);
```

Comment:
Why is the graph required to have inputs set?

Reply:
There was no setter and getter for these when I wrote it - @jacobkahn added it later - I can change it to use the new API